### DISCUSSION TITLE
How to pass $request_id to my application?

### DISCUSSION THREAD

**Discussion by txigreman** [nginx/unit#1183] [Support] [answered]
Hi,

I'm trying to make my application aware of the new variable `$request_id`, so I can use it in the logs and improve traceability.
I've tried to pass it in the `enviroment` section of the `application`, but njs templates does not seem to work there.

How could I achieve it?

Thanks in advance.

**Comment by ac000** [comment]
See [here](https://unit.nginx.org/configuration/#variables) for where and how these variables can be used.

**Comment by txigreman** [comment]
Thanks for your answer.

I read it before asking, and now I've read it again, but I still doesn't know how to achieve it. Any advice will be welcome.

**Comment by lcrilly** [comment, marked as answer]
I don't think it is possible to pass `$request_id` to an application without using `rewrite` to somehow embed it in the URI. You probably don't want to do that! :)

The conventional approach would be to decorate the request with a new `Request-ID: $request_id` header. Unit cannot do this and I cannot find a GitHub issue about it.

**Comment by txigreman** [comment]
Hi Icrilly,

I tried to pass `$request_id` in a custom header, as I was doing in Nginx, but I didn't find a proper way to achieve it. 

Yesterday, I found [this issue](https://github.com/nginx/unit/issues/1021) that you just renamed from "Support for request headers". Being able to set custom request headers would help to pass `$request_id` to the app, and other headers like X-Forwarded-Prefix, not mentioned in that issue.

In the same issue tippexs [suggested](https://github.com/nginx/unit/issues/1021#issuecomment-1852782491) to use environment variables as a workaround, that's why I tried to use them, but it only works for static values.

At this point, my idea was to add my use case to that issue, in support of adding the request headers feature; but now that is renamed I think it makes no sense.

Should I open a new issue? What do you think would be the best way to comunicate `$request_id` variable to the app? What makes more sense for me it's a custom header.

**Comment by lcrilly** [comment]
Please do create a new issue for the generic need. I was busy cleaning up to make space for it :)

**Comment by phpclub** [comment]
Variables not work in applications/*/environment section
```
"applications": {
    "php": {
      "environment": {
        "HTTP_X_REQUEST_ID": "$request_id",
        "MY_URI": "$uri"
      },
```

in PHP show as text

HTTP_X_REQUEST_ID	$request_id
MY_URI	$uri
